## <u style="color: blue">7. Gridsearch</u> 

Comme précisé dans le notebook principal, vous trouverez ici les calculs de Gridsearch pour définir le modèle le plus performant pour la prédiction de tweets concernant des catastrophes:
- Decision Tree
- Random Forest
- Linear Regression
- XGBoost
- SVM

In [1]:
# importation de toutes les librairies necessaires pour le gridsearch
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV

from sklearn.linear_model import LogisticRegression
from Decision_Tree import DecisionTree
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
import xgboost as xgb

In [2]:
# import du csv
df_train = pd.read_csv('./CSV/train_tweets_cleaned.csv')

In [3]:
df_train.head()

id                                               text  target  \
0   1                         Deeds Reason ALLAH Forgive       1   
1   4              Forest fire near La Ronge Sask Canada       1   
2   5  residents asked shelter place notified officer...       1   
3   6               receive evacuation orders California       1   
4   7                 sent photo Ruby smoke pours school       1   

                                    transformed_text  \
0            ['Deeds', 'Reason', 'ALLAH', 'Forgive']   
1  ['Forest', 'fire', 'near', 'La', 'Ronge', 'Sas...   
2  ['residents', 'asked', 'shelter', 'place', 'no...   
3  ['receive', 'evacuation', 'orders', 'California']   
4  ['sent', 'photo', 'Ruby', 'smoke', 'pours', 's...   

                                        stemmed_text  vectorized_text_TFIDF  \
0              ['deed', 'reason', 'allah', 'forgiv']               0.000169   
1  ['forest', 'fire', 'near', 'la', 'rong', 'sask...               0.000169   
2  ['resid', 'ask', 'shelter', 'place', 'notifi',...               0.000336   
3         ['receiv', 'evacu', 'order', 'california']               0.000631   
4  ['sent', 'photo', 'rubi', 'smoke', 'pour', 'sc...               0.000161   

                                 vectorized_text_BoW  
0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  
1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  
2  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  
3  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  
4  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...

In [4]:
X = df_train['vectorized_text_TFIDF']
y = df_train['target']

In [5]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y)

# XGBoost
# SVM
# Logistic Regression

Gridsearch pour le decision tree :

In [6]:
models_params = [
{
    'model': DecisionTreeClassifier(),
    'params': {
    'criterion': ['gini', 'entropy']
    },
},
{
    'model': RandomForestClassifier(),
    'params': {
    'criterion': ['gini', 'entropy']
    },
},
{
    'model': xgb.XGBClassifier(),
    'params': {
    },
},
{
    'model': svm.SVC(),
    'params': {
    'kernel': ['rbf', 'poly', 'sigmoid']
    },
},
{
    'model': LogisticRegression(),
    'params': {
    'solver': ['newton-cg', 'lbfgs', 'liblinear', 'sag', 'saga']
    },
}

]

In [10]:
scores = []

X_train = X_train.values.reshape(-1, 1)
y_train = y_train.values.reshape(-1, 1)

for model_param in models_params:
    clf = GridSearchCV(model_param['model'], model_param['params'], cv=5, n_jobs=-1, verbose=2)
    clf.fit(X_train, y_train)
    scores.append({
        'model': model_param['model'],
        'best_score': clf.best_score_,
        'best_params': clf.best_params_
    })

Fitting 5 folds for each of 2 candidates, totalling 10 fits
Fitting 5 folds for each of 2 candidates, totalling 10 fits


c:\Users\x_mat\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\base.py:1473: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)


Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 3 candidates, totalling 15 fits


c:\Users\x_mat\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\utils\validation.py:1310: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Fitting 5 folds for each of 5 candidates, totalling 25 fits
[{'model': DecisionTreeClassifier(), 'best_score': 0.5458622815960872, 'best_params': {'criterion': 'gini'}}, {'model': RandomForestClassifier(), 'best_score': 0.5370677906125294, 'best_params': {'criterion': 'gini'}}, {'model': XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=None, n_jobs=None,
              n

c:\Users\x_mat\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\utils\validation.py:1310: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [11]:
print(scores)

[{'model': DecisionTreeClassifier(), 'best_score': 0.5458622815960872, 'best_params': {'criterion': 'gini'}}, {'model': RandomForestClassifier(), 'best_score': 0.5370677906125294, 'best_params': {'criterion': 'gini'}}, {'model': XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=None, n_jobs=None,
              num_parallel_tree=None, random_state=None, ...), 'best_score'

[{'model': DecisionTreeClassifier(), 'best_score': 0.5458622815960872, 'best_params': {'criterion': 'gini'}}, {'model': RandomForestClassifier(), 'best_score': 0.5370677906125294, 'best_params': {'criterion': 'gini'}}, {'model': XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=None, n_jobs=None,
              num_parallel_tree=None, random_state=None, ...), 'best_score': 0.549980555176002, 'best_params': {}}, {'model': SVC(), 'best_score': 0.5930363005083683, 'best_params': {'kernel': 'poly'}}, {'model': LogisticRegression(), 'best_score': 0.5930363005083683, 'best_params': {'solver': 'newton-cg'}}]

Le meilleur modèle est le SVM avec un score de 0.5930363005083683 et les paramètres suivants : {'kernel': 'poly'}

In [16]:
# gridsearch for best hyperparameters for svM

model = svm.SVC()
params = {
    'kernel': ['rbf', 'poly', 'sigmoid'],
    'C': [0.1, 1, 10, 100, 1000],
    'gamma': [1, 0.1, 0.01, 0.001, 0.0001]
}

models_scores = []

clf = GridSearchCV(model, params, cv=5, n_jobs=-1, verbose=2)
clf.fit(X_train, y_train)
models_scores.append({
    'model': model,
    'best_score': clf.best_score_,
    'best_params': clf.best_params_
})

Fitting 5 folds for each of 75 candidates, totalling 375 fits


c:\Users\x_mat\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\utils\validation.py:1310: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [17]:
print(models_scores)

[{'model': SVC(), 'best_score': 0.5930363005083683, 'best_params': {'C': 0.1, 'gamma': 1, 'kernel': 'rbf'}}]


Maintenant que nous en avons terminé avec la partie du Gridsearch, nous pouvons retourner sur le premier notebook afin de mettre en application les résultats obtenus.

## <u style="color: blue">[8. Evaluation des modèles](./Eye_Of_Emergency.ipynb#8.-Evaluation-des-modèles)</u>